### PR TITLE
mac build fix

### PR DESCRIPTION
**Pull request by RodrigoCard**
This commit includes a bunch of fixes:
- gcc string problem workaround
- c++11 needed to build the game correctly (dont remember why)
- toString ambiguity fixed
- included a stub Joystick support for mac (so it could be built with
  games that expect joystick functions, but without joystick support, for
  now)

### CODE REVIEW

**Comment by RodrigoCard** [comment]
New fixes:

- Running from the IDE
- Creating a correct App Bundle in the choosen location in Compile Mode.

If there is nothing wrong I can call it done :)

File: CompilerSource/JDI/src/System/macros.cpp [hunk 1]
```diff
@@ -153,7 +153,7 @@ inline void jdip::macro_function::preparse(string val, error_handler *herr)
         value.push_back(mv_chunk(val.c_str(), push_from, i-push_from)); // Push it onto our value
       while (is_useless_macros(val[++i]));
       push_from = i; // Store current position just in case something stupid happens
-      if (!is_letter(val[i])) { herr->error("Expected parameter name following '#' token; `" + val + "'[" + ::toString(i) + "] is not a valid identifier character"); continue; }
+      if (!is_letter(val[i])) { herr->error("Expected parameter name following '#' token; `" + val + "'[" + ::toString((int)i) + "] is not a valid identifier character"); continue; }
```
Comment (JoshDreamland) [inline CompilerSource/JDI/src/System/macros.cpp]: I think someone has already fixed this error in a more general way. Is this change still needed?

Comment (RodrigoCard) [inline CompilerSource/JDI/src/System/macros.cpp]: Is the other fix on master branch already? If not where I can find it?
Because it simply doesnt work without it for me.

If you prefer, I can remove this one, and exchange for the other fix.

Comment (RobertBColton) [inline CompilerSource/JDI/src/System/macros.cpp]: I went and looked through all the branches on master for the file:
https://github.com/enigma-dev/enigma-dev/blob/master/CompilerSource/JDI/src/System/macros.cpp

I can't see anybody modifying except @JoshDreamland since 2012. Even on Harri's GL3.3 branches but I could have swore I seen this change already myself too.

Comment (RodrigoCard) [inline CompilerSource/JDI/src/System/macros.cpp]: I think this is the relevant change https://github.com/enigma-dev/enigma-dev/blob/GL3.3NormalMatrix/CompilerSource/JDI/src/General/parse_basics.cpp

**Comment by RodrigoCard** [comment]
After the last Merge I did ( 788c8b5 ) Something broke.

```
Operating System: Mac OS X
Version: 10.11.6
Architecture: x86_64

Java Name: Java HotSpot(TM) 64-Bit Server VM
Java Vendor: Oracle Corporation
Version: 1.7.0_21

Available processors (cores): 2
Free memory (bytes): 8594696
Maximum memory (bytes): 251396096
Total memory available to JVM (bytes): 30175232

File system root: /
Total space (bytes): 999210909696
Free space (bytes): 116789239808
Usable space (bytes): 116527095808

Stack trace:

java.lang.UnsatisfiedLinkError: Unable to load library 'compileEGMf': Native library (darwin/libcompileEGMf.dylib) not found in resource path ([file:/Users/Rodrigo/Workspace%20Libs/enigma-dev/plugins/enigma.jar])
	at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:271)
	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:398)
	at com.sun.jna.Library$Handler.<init>(Library.java:147)
	at com.sun.jna.Native.loadLibrary(Native.java:412)
	at com.sun.jna.Native.loadLibrary(Native.java:391)
	at org.enigma.EnigmaRunner.attemptLib(EnigmaRunner.java:248)
	at org.enigma.EnigmaRunner.access$1(EnigmaRunner.java:241)
	at org.enigma.EnigmaRunner$3.run(EnigmaRunner.java:179)
```
I already tried to put the dylib in "(root_of_project):/darwin/" folder, no success.
Where should I look for the error?

**Comment by RobertBColton** [comment]
Take a look at #1023 we pulled something that may have caused that.

**Comment by RodrigoCard** [comment]
Yeah! That was the culprit! java 64 bit library trying to load a 32 bit lib :)

**Comment by RobertBColton** [comment]
Thanks for letting me know, I've decided to merge that until we actually get a 64 bit release for Windows.

File: CompilerSource/JDI/src/System/macros.cpp (hunk 1 above)
Comment (JoshDreamland) [inline CompilerSource/JDI/src/System/macros.cpp]: Is this fixed, now? That is, can this file be excluded?

File: ENIGMAsystem/SHELL/Makefile [hunk 2]
```diff
@@ -59,14 +61,14 @@ CC := gcc
 DEPENDENCIES :=
 
 ifeq ($(GMODE), Debug)
-	override CXXFLAGS += -Wall -g -DDEBUG_MODE
+	override CXXFLAGS += -Wall -std=c++11 -g -DDEBUG_MODE
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Makefile]: I think you should be specifying these in the MacOSX `gcc.ey` rather than directly in the makefile.

File: ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+namespace enigma
+{
+  //stub
+}
+
+/* Note: empty stub functions to build Mac executables without errors (and no joystick support, of course) */
+namespace enigma_user
+{
+  bool joystick_load(int id) {
+    return false; //stub
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp]: Stubs are kind of tacky... is this necessary to get a game to compile?

File: CompilerSource/compiler/components/parse_and_link.cpp [hunk 4]
```diff
@@ -95,7 +95,7 @@ int lang_CPP::compile_parseAndLink(EnigmaStruct *es,parsed_script *scripts[], ve
       // Keep a parsed record of this timeline
       tline_lookup[es->timelines[i].name].push_back(tlines.back());
       parser_main(newcode, &tlines.back()->pev, script_names);
-      edbg << "Parsed `" << es->timelines[i].name <<", moment: " <<es->timelines[i].moments[j].stepNo << "': " << tlines.back()->obj.locals.size() << " locals, " << tlines.back()->obj.globals.size() << " globals" << flushl;
+      edbg << "Parsed `" << es->timelines[i].name <<", moment: " <<es->timelines[i].moments[j].stepNo << "': " << (int)(tlines.back()->obj.locals.size()) << " locals, " << (int)tlines.back()->obj.globals.size() << " globals" << flushl;
```
Comment (JoshDreamland) [inline CompilerSource/compiler/components/parse_and_link.cpp]: I am hoping this file can be rolled back, now, too. If not, please let me know and I will make sure that it can.

File: ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp (hunk 3 above)
Comment (RodrigoCard) [inline ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp]: Well, the game was building on windows/linux but not mac because of the lack of joystick support. Unless the user remove manually all the joystick code from the game I think it is necessary while there is no platform parity...  Write once, compile anywhere :)

By the way, I have some experience with Game Controllers on Mac, I can do that in the future.

Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp]: It's just sad that parts of the engine are forcibly compiled into your game that introduce dependencies on joysticks. It's a smell—this means that LTO cannot remove those functions from your executable. I dislike stubs because I'd prefer you get compile errors rather than have a game built that does nothing when you tell it to load a joystick. That's just frustrating.

Still, this problem has nothing to do with your pull request, so no action to take, here.

Comment (RodrigoCard) [inline ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp]: How about using a #warning directive to tell the user at build time about that ?

File: ENIGMAsystem/SHELL/Makefile (hunk 2 above)
Comment (RodrigoCard) [inline ENIGMAsystem/SHELL/Makefile]: done

File: ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp (hunk 3 above)
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaJoystick.cpp]: That would be nice, but unfortunately, that would just warn when this file was compiled (rather than when the method was used). We could do a `show_error` here to let them know that the joystick load is failing because it's unimplemented, but that's not great, either. Don't worry about it. This is a problem with ENIGMA's compiler and the setup of the system, not with this pull request. ☺

File: CompilerSource/compiler/components/parse_and_link.cpp (hunk 4 above)
Comment (RodrigoCard) [inline CompilerSource/compiler/components/parse_and_link.cpp]: When I remove these casts I get these errors:

```
compiler/components/parse_and_link.cpp:98:112: error: use of overloaded operator '<<' is
      ambiguous (with operand types 'ideprint' and 'size_type' (aka 'unsigned long'))
  ...edbg << "Parsed `" << es->timelines[i].name <<", moment: " <<es->timelines[i].moments[j].stepNo << "': " << (tlines.back()->obj.locals.size())...
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./backend/ideprint.h:29:13: note: candidate function
  ideprint &operator<< (int16_t x);
            ^
./backend/ideprint.h:30:13: note: candidate function
  ideprint &operator<< (int32_t x);
            ^
./backend/ideprint.h:31:13: note: candidate function
  ideprint &operator<< (int64_t x);
            ^
./backend/ideprint.h:32:13: note: candidate function
  ideprint &operator<< (uint16_t x);
            ^
./backend/ideprint.h:33:13: note: candidate function
  ideprint &operator<< (uint32_t x);
            ^
./backend/ideprint.h:34:13: note: candidate function
  ideprint &operator<< (uint64_t x);
            ^
./backend/ideprint.h:35:13: note: candidate function
  ideprint &operator<< (char x);
            ^
./backend/ideprint.h:36:13: note: candidate function
  ideprint &operator<< (unsigned char x);
            ^
./backend/ideprint.h:37:13: note: candidate function
  ideprint &operator<< (double x);
            ^
```
_(And the same error for macros.cpp)_

---
So, I tried to properly fix it without using these casts by adding a
`ideprint &operator<< (size_t x);`

**It fixed the parse_and_link file without the use of casts** :)

But not the macros.cpp :(
I got some linker errors:
```
Undefined symbols for architecture x86_64:
  "parse_bacics::visible::toString(unsigned long)", referenced from:
      jdip::macro_function::parse(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >&, char*&, char*&, jdip::token_t, jdi::error_handler*) const in macros.o
      jdip::macro_function::preparse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, jdi::error_handler*) in macros.o
ld: symbol(s) not found for architecture x86_64
```

Looks like a template problem.

Not sure how to fix, any tips?

Comment (RobertBColton) [inline CompilerSource/compiler/components/parse_and_link.cpp]: That actually should not be necessary because I already overloaded it for ``size_t`` before I thought.
https://github.com/enigma-dev/enigma-dev/commit/5b99151173b1dad9acb059f9e99a75974801d42b

Hrm... actually looks like I missed ``size_t`` for ideprint. Follow the changes I made in that file to overload ideprint for ``size_t``.

**Comment by JoshDreamland** [comment]
Hey Rodrigo,

Sorry for the holdup. We've committed a proper fix for those type problems, now (thanks to `std::to_string`, which should work correctly on Mac where our implementation did not).

We have a build bot running for OS X, now; our hope is that the rest of this change is unnecessary, now.

All the same, thank you for your effort!
